### PR TITLE
GPII-4428: Restoring icon spacing when restarting explorer.

### PR DESCRIPTION
**Pull request by stegru**
There appears to be a bug in SystemParametersInfo: when setting the icon spacing, it's adjusted to the current DPI, but getting the value isn't adjusted. When explorer starts (or stops?) it re-applies the setting (like `set(get())`), causing the wrong setting to be applied.

This fix gets the setting before restarting explorer, then re-applies it after explorer has started.

### CODE REVIEW

**Comment by kaspermarkus** [comment]
Yana has tested this issue and confirmed that it is actually working

**Comment by kaspermarkus** [comment]
Update to latest master ... once updated, @amb26 will merge

**Comment by stegru** [comment]
@amb26, it's ready.

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js [hunk 1]
```diff
@@ -414,6 +414,9 @@ windows.user32 = ffi.Library("user32", {
     // https://docs.microsoft.com/windows/win32/api/winuser/nf-winuser-setwindowpos
     "SetWindowPos": [
         t.BOOL, [ t.HANDLE, t.HANDLE, "int", "int", "int", "int", t.UINT ]
+    ],
+    "SetThreadDpiAwarenessContext": [
```
Comment (javihernandez) [inline gpii/node_modules/WindowsUtilities/WindowsUtilities.js:418]: Can you please add the link to the ref documentation? https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setthreaddpiawarenesscontext

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js [hunk 2]
```diff
@@ -2265,4 +2268,55 @@ fluid.defaults("gpii.windows.rm", {
...
+
+/**
+ * Sets the icon cell size.
+ * @param {Object} iconSpacing The icon cell size
+ * @param {Number} iconSpacing.x The width, in pixels.
+ * @param {Number} iconSpacing.y The height, in pixels.
+ * @param {Boolean} update True to broadcast an update.
+ */
+gpii.windows.setIconSpacing = function (iconSpacing, update) {
+
```
Comment (javihernandez) [inline gpii/node_modules/WindowsUtilities/WindowsUtilities.js:2305]: Please remove this blank line.

File: gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js [hunk 3]
```diff
@@ -564,3 +564,24 @@ jqUnit.test("Testing removeFiles", function () {
...
+    var iconSpacing = { x: 78, y: 80 };
+    gpii.windows.setIconSpacing(iconSpacing);
+    var newValue = gpii.windows.getIconSpacing();
+    jqUnit.assertDeepEq("newValue must the new value", iconSpacing, newValue);
+
+    // Set it back to the original value
+    gpii.windows.setIconSpacing(origValue, true);
+    var revertedValue = gpii.windows.getIconSpacing();
+    jqUnit.assertTrue("revertedValue must be the original value", origValue, revertedValue);
+
```
Comment (javihernandez) [inline gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js:585]: More blank lines

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js [hunk 4]
```diff
@@ -2265,4 +2268,55 @@ fluid.defaults("gpii.windows.rm", {
...
+        iconSpacing.x = intBuffer.deref();
+
+        gpii.windows.spi.systemParametersInfoWPtr.SystemParametersInfoW(
+            gpii.windows.spi.actions.SPI_ICONVERTICALSPACING, 0, intBuffer, 0);
+        iconSpacing.y = intBuffer.deref();
+
+    } finally {
+        gpii.windows.user32.SetThreadDpiAwarenessContext(oldDpiAwareness);
+    }
+
```
Comment (javihernandez) [inline gpii/node_modules/WindowsUtilities/WindowsUtilities.js:2292]: Extra blank lines

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js (hunk 1 above)
Comment (stegru) [reply gpii/node_modules/WindowsUtilities/WindowsUtilities.js:418]: done

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js (hunk 2 above)
Comment (stegru) [reply gpii/node_modules/WindowsUtilities/WindowsUtilities.js:2305]: removed

File: gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js (hunk 3 above)
Comment (stegru) [reply gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js:585]: gone

File: gpii/node_modules/WindowsUtilities/WindowsUtilities.js (hunk 4 above)
Comment (stegru) [reply gpii/node_modules/WindowsUtilities/WindowsUtilities.js:2292]: done

**Comment by amb26** [comment]
Failure in SPI settings handler swapMouseButtonsConfig:
````
    "gpii.windows.spiSettingsHandler": {
        "com.microsoft.windows.mouseSettings": [
            {
                "settings": {
                    "SwapMouseButtonsConfig": {
                        "value": 0,
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Weird network breakage:
````
doitclient.exe : doit: network error: No error
    + CategoryInfo          : NotSpecified: (doit: network error: No error:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
Build step 'Execute shell' marked build as failure
Sending e-mails to: ci@lists.gpii.net
Finished: FAILURE
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Again odd failure:
````
02:25:56.177:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win using default matchmaker" - Message: Error making request to /user/os_win/login: socket hang up
02:25:56.177:  jq: Source:     at pok (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.fail (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at ClientRequest.<anonymous> (\\vboxsvr\vagrant\node_modules\kettle\lib\test\KettleTestUtils.http.js:159:16)
    at ClientRequest.emit (events.js:182:13)
    at Socket.socketOnEnd (_http_client.js:425:9)
    at Socket.emit (events.js:187:15)
    at endReadableNT (_stream_readable.js:1092:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Particular run of bad luck recently:
````
15:40:12.771:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win using default matchmaker" - Message: Successful login message returned {"isError":true,"message":"Error occurred during login: Timed out waiting for the SPI call to complete (see GPII-2001)","ignoreUserErrors":false}
15:40:12.772:  jq: Expected: -1
15:40:12.775:  jq: Actual: -1
15:40:12.775:  jq: Source:     at Object.assertNotEquals (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:141:19)
    at gpii.test.loginRequestListen (\\vboxsvr\vagrant\node_modules\gpii-universal\gpii\node_modules\testing\src\TestUtils.js:176:12)
    at wrapped (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:320:26)
    at fire (\\vboxsvr\vagrant\node_modules\infusion\src\framework\core\js\Fluid.js:1612:40)
    at IncomingMessage.<anonymous> (\\vboxsvr\vagrant\node_modules\kettle\lib\test\KettleTestUtils.http.js:150:13)
    at IncomingMessage.emit (events.js:187:15)
    at endReadableNT (_stream_readable.js:1092:12)
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
For God's sake..... VM creation failure

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Perfection is achieved on the point of collapse - ok to test

**Comment by amb26** [comment]
Getting a lot of these suddenly - some kind of networking problem on the CI machine?
````
05:07:22.477:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win using default matchmaker" - Message: Error making request to /user/os_win/login: socket hang up
05:07:22.477:  jq: Source:     at pok (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.fail (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at ClientRequest.<anonymous> (\\vboxsvr\vagrant\node_modules\kettle\lib\test\KettleTestUtils.http.js:159:16)
    at ClientRequest.emit (events.js:182:13)
    at Socket.socketOnEnd (_http_client.js:425:9)
    at Socket.emit (events.js:187:15)
    at endReadableNT (_stream_readable.js:1092:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
05:07:23.851:  systemSettingsHandler return[{"settingID": "SystemSettings_Display_BlueLight_ManualToggleQuickAction", "isError": false, "errorMessage": null, "returnValue": false}]
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Socket hangup again:
````
02:41:45.731:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win using default matchmaker" - Message: Error making request to /user/os_win/login: socket hang up
02:41:45.731:  jq: Source:     at pok (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.fail (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at ClientRequest.<anonymous> (\\vboxsvr\vagrant\node_modules\kettle\lib\test\KettleTestUtils.http.js:159:16)
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
SPI timeout:
````
04:23:28.500:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win using default matchmaker" - Message: Successful login message returned {"isError":true,"message":"Error occurred during login: Timed out waiting for the SPI call to complete (see GPII-2001)","ignoreUserErrors":false}
04:23:28.500:  jq: Expected: -1
04:23:28.500:  jq: Actual: -1
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
ok to test

**Pull request merged**